File: testio.rs
use std::{
    fs::File,
    io::{Cursor, Read, Result, Write},
};

fn process_data<R: Read, W: Write>(source: &mut R, destination: &mut W) -> Result<()> {
    const BUF_SIZE: usize = 8192;
    let mut buffer = [0u8; BUF_SIZE];
    let mut bytes_read;
    loop {
        let mut total_bytes_read = 0;
        while total_bytes_read < BUF_SIZE {
            bytes_read = source.read(&mut buffer[total_bytes_read..])?;
            if bytes_read == 0 {
                if total_bytes_read > 0 {
                    for x in &mut buffer[..total_bytes_read] {
                        *x ^= 0x5a;
                    }
                    destination.write_all(&buffer[..total_bytes_read])?;
                    destination.flush()?;
                }
                return Ok(());
            }
            total_bytes_read += bytes_read;
        }
        for x in &mut buffer {
            *x ^= 0x5a;
        }
        destination.write_all(&buffer)?;
        destination.flush()?;
    }

    // If short reads are okay:
    // loop {
    //     bytes_read = source.read(&mut buffer)?;
    //     if bytes_read == 0 {
    //         break;
    //     }
    //     buffer[..bytes_read].iter_mut().for_each(|x| *x ^= 0x5a);
    //     destination.write_all(&buffer[..bytes_read])?;
    //     destination.flush()?;
    // }
    // Ok(())
}

fn main() {
    let mut src = File::open("input.bin").unwrap();
    let mut dst = File::create("output.bin").unwrap();
    process_data(&mut src, &mut dst).unwrap();
    println!("File written to successfully.");

    let in_memory = b"Hello this is some random text!";
    let mut byte_input = Cursor::new(in_memory);
    let mut byte_output = Cursor::new(Vec::<u8>::new());
    process_data(&mut byte_input, &mut byte_output).unwrap();
    println!("Bytes copied successfully.");
    let result = byte_output.into_inner();
    println!(
        "{}",
        result
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect::<String>()
    );

    // TcpStream already implements std::io::Read and std::io::Write
}

File: TestIO.cs
using System.Text;

class TestIO
{
    private static void ProcessData(Stream source, Stream destination)
    {
        const int BUF_SIZE = 8192;
        byte[] buffer = new byte[BUF_SIZE];
        int bytesRead;
        while (true)
        {
            int totalBytesRead = 0;
            while (totalBytesRead < BUF_SIZE)
            {
                bytesRead = source.Read(buffer, totalBytesRead, BUF_SIZE - totalBytesRead);
                if (bytesRead == 0)
                {
                    if (totalBytesRead > 0)
                    {
                        for (int i = 0; i < totalBytesRead; i++)
                        {
                            buffer[i] ^= 0x5A;
                        }
                        destination.Write(buffer, 0, totalBytesRead);
                        destination.Flush();
                    }
                    return;
                }
                totalBytesRead += bytesRead;
            }
            for (int i = 0; i < BUF_SIZE; i++)
            {
                buffer[i] ^= 0x5A;
            }
            destination.Write(buffer);
            destination.Flush();
        }

        // If short reads are okay:
        // while ((bytesRead = source.Read(buffer)) > 0)
        // {
        //     Enumerable.Range(0, bytesRead).ToList().ForEach(i => buffer[i] ^= 0x5A);
        //     destination.Write(buffer, 0, bytesRead);
        //     destination.Flush();
        // }
    }

    static void Main(string[] args)
    {
        using (FileStream src = new("input.bin", FileMode.Open, FileAccess.Read))
        using (FileStream dst = new("output.bin", FileMode.Create, FileAccess.Write))
        {
            ProcessData(src, dst);
            Console.WriteLine("File written to successfully.");
        }

        byte[] inMemory = Encoding.UTF8.GetBytes("Hello this is some random text!");
        using MemoryStream byteInput = new(inMemory);
        using MemoryStream byteOutput = new();
        ProcessData(byteInput, byteOutput);
        Console.WriteLine("Bytes copied successfully.");
        byte[] result = byteOutput.ToArray();
        Console.WriteLine(string.Join("", result.Select(b => $"{b:x2}")));

        // For a network socket, use client.getStream()
    }
}
